hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ec/reconstruction/ECReconstructionCoordinator.java

```diff
@@ -0,0 +1,417 @@
+                (int) (configuration.getStreamBufferMaxSize() / configuration
+                    .getStreamBufferSize()),
+                ByteStringConversion.createByteBufferConversion(false));
+        targetBlockStreams[i] =
+            new ECBlockOutputStream(blockLocationInfo.getBlockID(),
+                this.containerOperationClient.getXceiverClientManager(),
+                Pipeline.newBuilder().setId(PipelineID.valueOf(
+                    targetMap.get(toReconstructIndexes.get(i)).getUuid()))
+                    .setReplicationConfig(repConfig).setNodes(ImmutableList
+                    .of(targetMap.get(toReconstructIndexes.get(i))))
+                    .setState(Pipeline.PipelineState.CLOSED).build(),
+                bufferPool, configuration, blockLocationInfo.getToken());
+        bufs[i] = byteBufferPool.getBuffer(false, repConfig.getEcChunkSize());
+        // Make sure it's clean. Don't want to reuse the erroneously returned
+        // buffers from the pool.
+        bufs[i].clear();
+      }
+
+      sis.setRecoveryIndexes(toReconstructIndexes.stream().map(i -> (i - 1))
+          .collect(Collectors.toSet()));
+      long length = safeBlockGroupLength;
+      while (length > 0) {
+        int readLen = sis.recoverChunks(bufs);
+        // TODO: can be submitted in parallel
+        for (int i = 0; i < bufs.length; i++) {
+          targetBlockStreams[i].write(bufs[i]);
+          if (isFailed(targetBlockStreams[i],
+              targetBlockStreams[i].getCurrentChunkResponseFuture())) {
+            // If one chunk response failed, we should retry.
+            // Even after retries if it failed, we should declare the
+            // reconstruction as failed.
+            // For now, let's throw the exception.
+            throw new IOException(
+                "Chunk write failed at the new target node: "
+                    + targetBlockStreams[i].getDatanodeDetails()
+                    + ". Aborting the reconstruction process.");
+          }
+          bufs[i].clear();
+        }
+        length -= readLen;
+      }
+
+      for (int i = 0; i < targetBlockStreams.length; i++) {
+        try {
+          targetBlockStreams[i]
+              .executePutBlock(true, true, blockLocationInfo.getLength());
+          if (isFailed(targetBlockStreams[i],
+              targetBlockStreams[i].getCurrentPutBlkResponseFuture())) {
+            // If one chunk response failed, we should retry.
+            // Even after retries if it failed, we should declare the
+            // reconstruction as failed.
+            // For now, let's throw the exception.
+            throw new IOException(
+                "Chunk write failed at the new target node: "
+                    + targetBlockStreams[i].getDatanodeDetails()
+                    + ". Aborting the reconstruction process.");
+          }
+        } finally {
+          byteBufferPool.putBuffer(bufs[i]);
+          targetBlockStreams[i].close();
```

Comment:
Loop would exit on first failed stream.  Seems like cleanup is not performed on the remaining ones.

Reply:
Yeah, good catch.